### ISSUE TITLE
Overhead memory usage when using merge with drain

### ISSUE THREAD

**Issue by Alexey-Yuferov** [typelevel/fs2#2607]
```scala
val src = IO(Array.ofDim[Int](1024))
(Stream.repeatEval(src) merge Stream.repeatEval(src)).drain.compile.drain.as(ExitCode.Success)
```
This code uses up all memory and CPU.
![image](https://user-images.githubusercontent.com/68867195/133265798-46905000-6b5d-42b1-bdba-33a15101b802.png)
The probem emerges only when there's a `drain` before `compile`. 
If we remove `drain` before `compile`, the example works normally, i.e:
```scala
val src = IO(Array.ofDim[Int](1024))
(Stream.repeatEval(src) merge Stream.repeatEval(src)).compile.drain.as(ExitCode.Success)
```
![image](https://user-images.githubusercontent.com/68867195/133265860-9ce687f4-1936-4d2e-ba50-0ca3e7cd0d6b.png)
out app executes multiple stream pipelines with `parJoinUnbounded`, and before the final compilation we have to drain each of those.

fs2 3.1.1 / cats-effect 3.2.3 / Scala 2.13.6

**Comment by mpilquist** [comment]
I think this is the memory leak that was just fixed in 3.1.2. Could you try again with that version?

**Comment by Alexey-Yuferov** [comment]
Yes, it works fine with 3.1.2, thank you
